### ISSUE TITLE
script goal does not work with additional dependencies

### ISSUE THREAD

**Issue by lefou** [davidB/scala-maven-plugin#166]
I tried to run a Scala script in non-Scala project (to replace an antrun target).

The project has no scala dependencies.

Here is the plugin configuration:
```xml
<plugin>
                                <groupId>net.alchim31.maven</groupId>
                                <artifactId>scala-maven-plugin</artifactId>
                                <version>3.2.0</version>
                                <executions>
                                        <execution>
                                                <id>test-scala</id>
                                                <phase>package</phase>
                                                <goals><goal>script</goal></goals>
                                                <configuration>
                                                        <scalaVersion>2.11.6</scalaVersion>
                                                        <dependencies>
                                                                <dependency>
                                                                        <groupId>com.lihaoyi</groupId>
                                                                        <artifactId>ammonite-ops_2.11</artifactId>
                                                                        <version>0.2.9</version>
                                                                </dependency>
                                                        </dependencies>
                                                        <script>
                                                                import java.io.File
                                                                println("1:" + new File("target").getAbsoluteFile())
                                                                import ammonite.ops._
                                                                println("2: " + cwd/'target)
                                                        </script>
                                                </configuration>
                                        </execution>
                                </executions>
                        </plugin>
```

And here is the error message:
```
[ERROR] Failed to execute goal net.alchim31.maven:scala-maven-plugin:3.2.0:script (test-scala) on project org.example.test: wrap: java.lang.Exception: A NoClass
DefFoundError exception was thrown: ammonite/ops/package$: ammonite.ops.package$ -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal net.alchim31.maven:scala-maven-plugin:3.2.0:script (test-scala) on project org.example.test: wrap: java.lang.Exception: A NoClassDefFoundError exception was thrown
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:216)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:862)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:286)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:197)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoExecutionException: wrap: java.lang.Exception: A NoClassDefFoundError exception was thrown
        at scala_maven.ScalaMojoSupport.execute(ScalaMojoSupport.java:490)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
        ... 20 more
Caused by: java.lang.Exception: A NoClassDefFoundError exception was thrown
        at scala_maven.ScalaScriptMojo.runScript(ScalaScriptMojo.java:224)
        at scala_maven.ScalaScriptMojo.doExecute(ScalaScriptMojo.java:171)
        at scala_maven.ScalaMojoSupport.execute(ScalaMojoSupport.java:482)
        ... 22 more
Caused by: java.lang.NoClassDefFoundError: ammonite/ops/package$
        at embeddedScript_1.<init>(embeddedScript_1.scala:10)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
        at scala_maven.ScalaScriptMojo.runScript(ScalaScriptMojo.java:190)
        ... 24 more
Caused by: java.lang.ClassNotFoundException: ammonite.ops.package$
        at org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy.loadClass(SelfFirstStrategy.java:50)
        at org.codehaus.plexus.classworlds.realm.ClassRealm.unsynchronizedLoadClass(ClassRealm.java:271)
        at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:247)
        at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:239)
        ... 30 more
```
The script runs successful until the import which is provided by the addition dependency. I can see the output of the first `println("1:")`, but after that, I get the exception. So, the issue is not a missing scala library.

Any ideas? What else do you need to debug?

**Comment by davidB** [comment]
I guess its a bug in the doc generation. dependencies should be added as sibbling of configuration, not as child. (see http://maven.apache.org/ref/3.3.1//maven-model/maven.html#class_plugin)

**Comment by lefou** [comment]
Unfortunately, moving the dependency into the plugin block doesn't solve the issue. The Error message is a little bit different though:

```
[INFO] --- scala-maven-plugin:3.2.0:script (test-scala) @ test ---
/tmp/test/target/.scalaScriptGen/embeddedScript_1.scala:4: error: not found: value ammonite
                                   import ammonite.ops._
                                          ^
/tmp/test/target/.scalaScriptGen/embeddedScript_1.scala:5: error: not found: value cwd
                                   println("2: " + cwd/'target)
                                                   ^
```

Here's the stack trace:

```
[ERROR] Failed to execute goal net.alchim31.maven:scala-maven-plugin:3.2.0:script (test-scala) on project test: wrap: org.apache.commons.exec.ExecuteException: Process exited with an error: 1 (Exit value: 1) -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal net.alchim31.maven:scala-maven-plugin:3.2.0:script (test-scala) on project test: wrap: org.apache.commons.exec.ExecuteException: Process exited with an error: 1 (Exit value: 1)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:216)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:862)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:286)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:197)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoExecutionException: wrap: org.apache.commons.exec.ExecuteException: Process exited with an error: 1 (Exit value: 1)
        at scala_maven.ScalaMojoSupport.execute(ScalaMojoSupport.java:490)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
        ... 20 more
Caused by: org.apache.commons.exec.ExecuteException: Process exited with an error: 1 (Exit value: 1)
        at org.apache.commons.exec.DefaultExecutor.executeInternal(DefaultExecutor.java:377)
        at org.apache.commons.exec.DefaultExecutor.execute(DefaultExecutor.java:160)
        at org.apache.commons.exec.DefaultExecutor.execute(DefaultExecutor.java:147)
        at scala_maven_executions.JavaMainCallerByFork.run(JavaMainCallerByFork.java:100)
        at scala_maven_executions.JavaMainCallerSupport.run(JavaMainCallerSupport.java:110)
        at scala_maven.ScalaScriptMojo.compileScript(ScalaScriptMojo.java:256)
        at scala_maven.ScalaScriptMojo.doExecute(ScalaScriptMojo.java:170)
        at scala_maven.ScalaMojoSupport.execute(ScalaMojoSupport.java:482)
        ... 22 more

```

**Comment by davidB** [comment]
FYI, I found the bug and fix it. But I can't release new version, due to some change in maven api that break integration-test. Upgrade the code to work with maven 3.2.+ will need more time.

Sorry.

**Comment by lefou** [comment]
Great, thank you!

I'm just curious. Do you plan to build against Maven 3.2.+ API? Does that mean, that next release will not work with 3.1?

**Comment by davidB** [comment]
Yes, I plan to require maven 3.2.+. is it a problem ?

**Comment by lefou** [comment]
Not for me. What's the reason? Any bugs or features what won't work with 3.0+? My feeling is, that abandoning < 3 is OK, but 3.0+ should work.

**Comment by davidB** [comment]
My integration test no longer pass when running with maven 3.2+. (use too
many deprecated api, the plugin was started 7 ago ;-) )

**Comment by lefou** [comment]
Hmm, I know what you mean. It's hard to come with a reasonable version set of all those Maven artifacts needed to build a plugin except the one you got when you first created the project.

No other objections beside the fear to exclude way to many user.

**Comment by davidB** [comment]
3.2.1 available in central repo in 24H

**Comment by danieldietrich** [comment]
Yay!

**Comment by lefou** [comment]
Giving the required dependencies in the plugin-block now works, great. Thank you!

Although, I'm wondering, why the initial snippet with the dependencies-block in the configuration part also forces Maven to download the given dependency?

**Comment by davidB** [comment]
I don't understand. Maven should download only missing dependencies (into local cache). Can you provide more info ?
Note: you could use "includeScopes" to use dependencies define in scope : test,compile,runtime.

**Comment by davidB** [comment]
Sorry for delay.

**Comment by lefou** [comment]
In the example I gave in the issue description, I declared an additional dependency to the configuration-block. Later you told me, this is wrong and not supported and just a glitch in the documentation generator, which is all perfectly fine for me. But still, I find it a bit strange, that although that said example it is not supported and it does not work as intended, Maven downloads the additional dependency. It's not an issue. I just wanted to tell you.
